test/integration/good/warning/pretty.expected

```diff
@@ -563,6 +563,53 @@ Warning in 'overloaded_forward_decl.stan', line 2, column 6: Functions do not
 Warning in 'overloaded_forward_decl.stan', line 3, column 6: Functions do not
     need to be declared before definition; all user defined function names
     are always in scope regardless of definition order.
+  $ ../../../../../install/default/bin/stanc --auto-format self-assign.stan
+model {
+  real foo = 1.5;
+  // line 4: should warn
+  foo = foo;
+  // line 6: should warn (decl)
+  array[2] real bar = bar;
+  // line 8: should warn (decl)
+  real baz = foo + baz;
+  // no warnings
+  bar[2] = bar[1];
+  //line 12: should warn
+  bar[1] = bar[1];
+  
+  // line 15: should warn
+  foo = (foo);
+  // no warnings
+  foo = -foo;
+  foo = (foo > 0) ? foo : 0.0;
+  
+  tuple(real, real) tup = (1.0, 2.0);
+  // line 22: should warn
+  tup.1 = tup.1;
+  // line 24: should warn (decl)
+  tuple(real, real) tup2 = (tup2.2, tup2.2);
```

Comment:
```suggestion
  // line 24: should warn (decl)
  tuple(real, real) tup2 = (tup2.1, tup2.2);
```
Also let's add a test for this "tuple reconstruction" case outside of declaration. It's similar to line 27 (could warn, currently does not)

Reply:
I think we should warn if any element of a tuple is accessed during declaration, so even something like this:

`tuple(real, real) x = (0.5, x.1);`

should warn, even though there is no memory address being directly assigned to itself in the end. 

Not during a declaration, I think we’d only want to warn if the same element is being assigned to itself (which this PR as currently written wouldn’t do, but I’m happy to write up the tests)

Reply:
> we’d only want to warn if the same element is being assigned to itself (which this PR as currently written wouldn’t do

just like the "line 27" test.

> we should warn if any element of a tuple is accessed during declaration

yes, but the test as written confusingly _looks_ like a typo.
In all honesty, I'd rather have `--warn-uninitialized` enabled by default in stead of this special-cased "used while declared" warning.

Reply:
`--warn-uninitialized` is run in the MIR layers, so in particular it does not produce any output when using something like `--auto-format` which exits before lowering to MIR.

It also looks like it warns when using a higher-order function that the functor argument is uninitialized, which is a bug.

There's some redundancy between the two but I think that's fine if we know this subset is correct

Reply:
I know `--warn-uninitialized` is broken. I just wish less effort was spent on reimplementing it in AST and more on fixing it.

Reply:
I tried enabling `--warn-uninitialized` by default and it caused the tests to take _significantly_ longer to run. There are a few obvious improvements we could make to it (like fixing the higher-order function issue) but it also seems to have some more underlying performance issue we'd need to sort out